### PR TITLE
Fix E2E tests

### PR DESCRIPTION
**Pull request by gigitux**
This PR improves the stability of our E2E tests.

In this PR:
- some E2E tests were updated to improve the flaky status.
- increased the timeout of `jest-puppeteer`.
- upgraded `@wordpress/e2e-tests"` and `jest-circus` packages
- Disabled `exitOnPageError` option.

####  Disabled `exitOnPageError` option.

This option was disabled to avoid the error:

```
Expected an Error, but "" was thrown. 
```

Sometimes the `JQuery minified file` (from WordPress) throws an error. By default, Puppeteer exits if a script throws an error. `exitOnPageError` option disabled this behavior. 

#### User Facing Testing

Run the following command and be sure that there aren't any errors from the E2E folder.

```
npm run test:e2e
```

* [x] Do not include in the Testing Notes

### CODE REVIEW

File: tests/e2e/config/environment.js [hunk 1]
```diff
@@ -23,4 +23,36 @@ class E2EEnvironment extends PuppeteerEnvironment {
 	}
 }
 
+// This code is useful for debug purposes.
```
Comment (sunyatasattva) [inline tests/e2e/config/environment.js]: I assume you left this comments in on purpose, so that people could easily uncomment it to debug stuff, right?

If that's the case, could you write a bit more on the comment about why and how it is useful and how and when should people use this code?

File: tests/e2e/config/jest.setup.js [hunk 2]
```diff
@@ -8,10 +9,17 @@ import {
 	switchUserToAdmin,
 	switchUserToTest,
 	visitAdminPage,
+	clearLocalStorage,
 } from '@wordpress/e2e-test-utils';
+/**
+ * Internal dependencies
+ */
+import { DEFAULT_TIMEOUT } from '../utils';
+const setDefaultOptions = require( 'expect-puppeteer' ).setDefaultOptions;
```
Comment (sunyatasattva) [inline tests/e2e/config/jest.setup.js]: Sorry, maybe that's a stupid question: can't it be imported normally? Why does it need to be `require`d? Maybe since this is odd, let's put a comment on it?
Can we at least destructure it, you think?

File: tests/e2e/specs/shopper/active-filters.test.ts [hunk 3]
```diff
@@ -1,3 +1,4 @@
+/* eslint-disable import/no-extraneous-dependencies */
```
Comment (sunyatasattva) [inline tests/e2e/specs/shopper/active-filters.test.ts]: Why is this needed? Because Puppeteer isn't in our `devDependencies`?

File: tests/e2e/specs/shopper/active-filters.test.ts [hunk 4]
```diff
@@ -36,6 +34,8 @@ const block = {
 		frontend: {
 			activeFilterType: '.wc-block-active-filters__list-item-type',
 			activeFilterName: '.wc-block-active-filters__list-item-name',
+			inStockInput: "label[for='instock']",
+			'128gbInput': "label[for='128gb']",
```
Comment (sunyatasattva) [inline tests/e2e/specs/shopper/active-filters.test.ts]: I don't see those selectors used anywhere, maybe you didn't need them anymore?

As a suggestion, maybe let's use a selector factory here, something like:

```suggestion
			filterCheckboxFor( label: string ) => `label[for="${ label }"]`
```

File: tests/e2e/specs/shopper/active-filters.test.ts [hunk 5]
```diff
@@ -113,18 +113,21 @@ describe( 'Shopper → Active Filters Block', () => {
 				}
 			);
 
+			await page.waitForSelector(
+				selectors.frontend.productsList + '.is-loading',
```
Comment (sunyatasattva) [inline tests/e2e/specs/shopper/active-filters.test.ts]: Just curious, why did you use this method of concatenation over template literal?

File: tests/e2e/specs/shopper/active-filters.test.ts [hunk 6]
```diff
@@ -113,18 +113,21 @@ describe( 'Shopper → Active Filters Block', () => {
...
+
 			await expect( page ).toClick( 'label', {
 				text: 'In stock',
 			} );
 
-			await page.waitForSelector( block.class );
+			await page.screenshot( {
+				path: 'aaa.png',
+				fullPage: true,
+			} );
```
Comment (sunyatasattva) [inline tests/e2e/specs/shopper/active-filters.test.ts]: This has probably been forgotten, I assume? 🤔

File: tests/e2e/specs/shopper/active-filters.test.ts [hunk 7]
```diff
@@ -113,18 +113,21 @@ describe( 'Shopper → Active Filters Block', () => {
...
 
-			await expect( page ).toMatchElement(
-				selectors.frontend.activeFilterType,
-				{
-					text: 'Stock Status',
-				}
-			);
+			await expect( page ).toMatch( 'Stock Status' );
```
Comment (sunyatasattva) [inline tests/e2e/specs/shopper/active-filters.test.ts]: I like that you have refactored the result expectation out (although I probably wouldn't have put it as a property of the `block` object, as it seems a bit confusing), what do you think about getting this string out as well? I'm not sure what it is, but something like `FILTER_TITLE` ?

File: tests/e2e/specs/shopper/active-filters.test.ts [hunk 8]
```diff
@@ -144,7 +147,7 @@ describe( 'Shopper → Active Filters Block', () => {
 			await expect( page ).toMatch( block.foundProduct );
 		} );
 
-		fit( 'When clicking the X on a filter it removes a filter', async () => {
+		it( 'When clicking the X on a filter it removes a filter', async () => {
```
Comment (sunyatasattva) [inline tests/e2e/specs/shopper/active-filters.test.ts:148]: Hehe fit! 💪

File: tests/e2e/specs/shopper/cart-checkout/account.test.js [hunk 9]
```diff
@@ -15,7 +15,7 @@ import { visitAdminPage } from '@wordpress/e2e-test-utils';
 /**
  * Internal dependencies
  */
-import { shopper, merchant } from '../../../../utils';
+import { shopper, merchant, clickLink } from '../../../../utils';
 import { SIMPLE_PHYSICAL_PRODUCT_NAME } from '.../../../../utils/constants';
```
Comment (sunyatasattva) [inline tests/e2e/specs/shopper/cart-checkout/account.test.js:19]: Just noticed this!

Regarding my comment above, this is a much better way to put the string instead of inside the `block` object IMHO. Also, it's literally the same string as your `foundProduct`, so what do you say about using this constant instead of the `foundProduct` above?

File: tests/e2e/specs/shopper/filter-products-by-price.test.ts [hunk 10]
```diff
@@ -59,7 +59,7 @@ const setMaxPrice = async () => {
 	await page.click( selectors.frontend.priceMaxAmount, {
 		clickCount: 3,
 	} );
-	await page.keyboard.type( '1.99' );
+	await page.keyboard.type( '2' );
```
Comment (sunyatasattva) [inline tests/e2e/specs/shopper/filter-products-by-price.test.ts]: Haha that's interesting. Why this change?

File: tests/e2e/specs/shopper/filter-products-by-price.test.ts [hunk 11]
```diff
@@ -97,10 +97,10 @@ describe( `${ block.name } Block`, () => {
 			await setMaxPrice();
 			await page.waitForNetworkIdle();
 			await waitForAllProductsBlockLoaded();
-			const products = await page.$$( selectors.frontend.productsList );
+
+			await page.waitForTimeout( 1000 );
```
Comment (sunyatasattva) [inline tests/e2e/specs/shopper/filter-products-by-price.test.ts]: Rogue hard coded timeout, probably forgotten?

File: tests/e2e/specs/shopper/filter-products-by-price.test.ts [hunk 12]
```diff
@@ -196,14 +191,10 @@ describe( `${ block.name } Block`, () => {
...
 			await clickLink( selectors.frontend.submitButton );
 
-			const products = await page.$$(
-				selectors.frontend.classicProductsList
-			);
 			const pageURL = page.url();
 			const parsedURL = new URL( pageURL );
 
 			expect( isRefreshed ).toBeCalledTimes( 1 );
-			expect( products ).toHaveLength( 1 );
```
Comment (sunyatasattva) [inline tests/e2e/specs/shopper/filter-products-by-price.test.ts]: Why are we removing this check?

**Review by sunyatasattva — CHANGES_REQUESTED** [review]
Added a few comments. Code looks good!

File: tests/e2e/specs/shopper/filter-products-by-price.test.ts (hunk 12 above)
Comment (gigitux) [reply tests/e2e/specs/shopper/filter-products-by-price.test.ts]: Re-added 👍

File: tests/e2e/specs/shopper/filter-products-by-price.test.ts (hunk 11 above)
Comment (gigitux) [reply tests/e2e/specs/shopper/filter-products-by-price.test.ts]: Removed 👍

File: tests/e2e/specs/shopper/filter-products-by-price.test.ts (hunk 10 above)
Comment (gigitux) [reply tests/e2e/specs/shopper/filter-products-by-price.test.ts]: Just for debug reason, I restored the change.

File: tests/e2e/specs/shopper/active-filters.test.ts (hunk 6 above)
Comment (gigitux) [reply tests/e2e/specs/shopper/active-filters.test.ts]: ahahah, yes! Powerful debugger :P

File: tests/e2e/specs/shopper/active-filters.test.ts (hunk 5 above)
Comment (gigitux) [reply tests/e2e/specs/shopper/active-filters.test.ts]: Just for velocity. I will update the code

File: tests/e2e/specs/shopper/active-filters.test.ts (hunk 4 above)
Comment (gigitux) [reply tests/e2e/specs/shopper/active-filters.test.ts]: Removed the not used selectors 👍

File: tests/e2e/config/jest.setup.js (hunk 2 above)
Comment (gigitux) [reply tests/e2e/config/jest.setup.js]: Not a stupid question :D 

I copied and paste the import, I will update it 👍

File: tests/e2e/specs/shopper/active-filters.test.ts (hunk 7 above)
Comment (gigitux) [reply tests/e2e/specs/shopper/active-filters.test.ts]: Thanks for pointing this out. [I extracted more string](https://github.com/woocommerce/woocommerce-gutenberg-products-block/blob/42e552d5f28c64f6e675f5a0a3eb145d934a3088/tests/e2e/specs/shopper/active-filters.test.ts/#L51-L54). Do you think that we should put these constants in `utils/constants.js`. I think that for these one it doesn't make sense because they are not for general purpose, but just for this block.

**Comment by gigitux** [comment]
> Added a few comments. Code looks good!

Thanks for your great review! I address your feedback. I extract[ `waitForAllProductsBlockLoaded` function]( https://github.com/woocommerce/woocommerce-blocks/blob/9895da816b3fb04ed0b567c48df3c03689c5b433/tests/e2e/utils.js#L451-L456) and moved in the utils file.

I guess that we should start to split the utils.js file. It has become huge.

File: tests/e2e/specs/shopper/active-filters.test.ts (hunk 3 above)
Comment (gigitux) [reply tests/e2e/specs/shopper/active-filters.test.ts]: Puppeteer isn't in our devDependencies because it is installed by `wordpress/e2e-tests` package.

Comment (sunyatasattva) [reply tests/e2e/specs/shopper/active-filters.test.ts]: Mm, I think that as long as Puppeteer is a peer dependency of another package, that's fine. But since we are now importing something from it, I believe this should be added to the `devDependencies`, what do you think?

Like, if `wordpress/e2e-tests` dependencies change (even in version) this might cause errors in this code. Since we are importing directly, I think it now becomes an internal dep.

File: tests/e2e/specs/shopper/active-filters.test.ts (hunk 7 above)
Comment (sunyatasattva) [reply tests/e2e/specs/shopper/active-filters.test.ts]: I agree with you. If they are only for this block, it's ok to have them scoped in this file. If you can see some other test reusing them, it'd be nice to have them in the constants file.

File: tests/e2e/utils.js [hunk 13]
```diff
@@ -448,14 +448,10 @@ export const openBlockEditorSettings = async () => {
...
- * Click a link and wait for the page to load.
- *
- * @param {string} selector The CSS selector of the link to click.
- */
-export const clickLink = async ( selector ) => {
-	await Promise.all( [
-		page.click( selector ),
-		page.waitForNavigation( { waitUntil: 'networkidle0' } ),
-	] );
+export const waitForAllProductsBlockLoaded = async () => {
```
Comment (sunyatasattva) [inline tests/e2e/utils.js:457]: Small nitpick, not blocking:

1. Could you add a little JSDoc to this? I know the name is self explanatory, but it might be worth explaining what it does and it also encourages contributors to always put JSDocs on their functions.
2. I know a lot of selectors here are hard-coded, but I would like to see it extracted somewhere else, if you have the patience for that 😊

**Review by sunyatasattva — APPROVED** [review]
I'm pre-approving, but I'd like to see Puppeteer being added to `devDependencies`. The other change is optional.

Great job and thanks for addressing all my issues and creating clean code!

I agree we should start thinking of splitting our utils file!

**Comment by gigitux** [comment]
> I'm pre-approving, but I'd like to see Puppeteer being added to `devDependencies`. The other change is optional.
> 
> Great job and thanks for addressing all my issues and creating clean code!
> 
> I agree we should start thinking of splitting our utils file!

I addressed all your suggestions! Thanks again! 🎉
